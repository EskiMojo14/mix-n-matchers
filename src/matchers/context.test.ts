/* eslint-disable @typescript-eslint/no-empty-function */
import type { Mock } from "@globals";
import { it, expect, describe, fn } from "@globals";
import type { CallInfo, Spy } from "../utils";

// Given a Jest mock function, return a minimal mock of a spy.
const createSpy = (mock: Mock): Spy => {
  const spy = function () {};

  spy.calls = {
    all() {
      const info: Array<CallInfo> = [];
      let i = 0;
      while (i < mock.mock.calls.length) {
        const returnRecord = mock.mock.results[i];
        /* eslint-disable @typescript-eslint/no-unsafe-assignment */
        info.push({
          args: mock.mock.calls[i] ?? [],
          object:
            "contexts" in mock.mock
              ? mock.mock.contexts[i]
              : (mock.mock as { instances: Array<unknown> }).instances[i],
          returnValue:
            returnRecord?.type === "return" ? returnRecord.value : undefined,
        });
        /* eslint-enable @typescript-eslint/no-unsafe-assignment */
        i++;
      }
      return info;
    },
    count() {
      return mock.mock.calls.length;
    },
  };

  return spy;
};

describe.each([
  "toHaveBeenCalledWithContext",
  "toHaveBeenLastCalledWithContext",
  "toHaveBeenNthCalledWithContext",
] as const)("%s", (calledWithContext) => {
  it("works only on spies or fn", () => {
    const mock = function mock() {};
    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(() => {
        expect(mock)[calledWithContext](3, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(() => {
        expect(mock)[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works when not called", () => {
    const mock = fn();
    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(createSpy(mock)).not[calledWithContext](1, "foo");
      expect(mock).not[calledWithContext](1, "foo");

      expect(() => {
        expect(mock)[calledWithContext](1, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(createSpy(mock)).not[calledWithContext]("foo");
      expect(mock).not[calledWithContext]("foo");

      expect(() => {
        expect(mock)[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works with a context that doesn't match", () => {
    const mock = fn();
    mock.call("bar");

    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(createSpy(mock)).not[calledWithContext](1, "foo");
      expect(mock).not[calledWithContext](1, "foo");

      expect(() => {
        expect(mock)[calledWithContext](1, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(createSpy(mock)).not[calledWithContext]("foo");
      expect(mock).not[calledWithContext]("foo");

      expect(() => {
        expect(mock)[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works with a context that doesn't match with matchers", () => {
    const mock = fn();
    mock.call("bar");

    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(createSpy(mock)).not[calledWithContext](1, expect.any(Number));
      expect(mock).not[calledWithContext](1, expect.any(Number));

      expect(() => {
        expect(mock)[calledWithContext](1, expect.any(Number));
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(createSpy(mock)).not[calledWithContext](expect.any(Number));
      expect(mock).not[calledWithContext](expect.any(Number));

      expect(() => {
        expect(mock)[calledWithContext](expect.any(Number));
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works with a context that matches", () => {
    const mock = fn();
    mock.call("foo");

    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(createSpy(mock))[calledWithContext](1, "foo");
      expect(mock)[calledWithContext](1, "foo");

      expect(() => {
        expect(mock).not[calledWithContext](1, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(createSpy(mock))[calledWithContext]("foo");
      expect(mock)[calledWithContext]("foo");

      expect(() => {
        expect(mock).not[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
  it("works with a context that matches with a matcher", () => {
    const mock = fn();
    mock.call("foo");

    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(createSpy(mock))[calledWithContext](1, expect.any(String));
      expect(mock)[calledWithContext](1, expect.any(String));

      expect(() => {
        expect(mock).not[calledWithContext](1, expect.any(String));
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(createSpy(mock))[calledWithContext](expect.any(String));
      expect(mock)[calledWithContext](expect.any(String));

      expect(() => {
        expect(mock).not[calledWithContext](expect.any(String));
      }).toThrowErrorMatchingSnapshot();
    }
  });

  it("includes the custom mock name in the error message", () => {
    const mock = fn().mockName("named-mock");

    mock.call("foo");

    if (calledWithContext === "toHaveBeenNthCalledWithContext") {
      expect(mock)[calledWithContext](1, "foo");

      expect(() => {
        expect(mock).not[calledWithContext](1, "foo");
      }).toThrowErrorMatchingSnapshot();
    } else {
      expect(mock)[calledWithContext]("foo");

      expect(() => {
        expect(mock).not[calledWithContext]("foo");
      }).toThrowErrorMatchingSnapshot();
    }
  });
});
