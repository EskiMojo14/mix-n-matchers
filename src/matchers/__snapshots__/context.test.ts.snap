// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toHaveBeenCalledWithContext includes the custom mock name in the error message 1`] = `
<d>expect(</b><r>named-mock</g><d>).</b>not<d>.</b>toHaveBeenCalledWithContext<d>(</b><g>expected</g><d>)</b>

Expected: not <g>"foo"</g>
Received:     <d>"foo"</b>

Number of calls: <r>1</g>
`;

exports[`toHaveBeenCalledWithContext works only on spies or fn 1`] = `
<d>expect(</b><r>received</g><d>).</b>toHaveBeenCalledWithContext<d>(</b><g>expected</g><d>)</b>

<b>Matcher error</b>: <r>received</g> value must be a mock or spy function

Received has type:  function
Received has value: <r>[Function mock]</g>
`;

exports[`toHaveBeenCalledWithContext works when not called 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>toHaveBeenCalledWithContext<d>(</b><g>expected</g><d>)</b>

Expected: <g>"foo"</g>

Number of calls: <r>0</g>
`;

exports[`toHaveBeenCalledWithContext works with a context that doesn't match 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>toHaveBeenCalledWithContext<d>(</b><g>expected</g><d>)</b>

Expected: <g>"foo"</g>
Received: <r>"bar"</g>
Number of calls: <r>1</g>
`;

exports[`toHaveBeenCalledWithContext works with a context that doesn't match with matchers 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>toHaveBeenCalledWithContext<d>(</b><g>expected</g><d>)</b>

Expected: <g>Any<Number></g>
Received: <r>"bar"</g>
Number of calls: <r>1</g>
`;

exports[`toHaveBeenCalledWithContext works with a context that matches 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>not<d>.</b>toHaveBeenCalledWithContext<d>(</b><g>expected</g><d>)</b>

Expected: not <g>"foo"</g>
Received:     <d>"foo"</b>

Number of calls: <r>1</g>
`;

exports[`toHaveBeenCalledWithContext works with a context that matches with a matcher 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>not<d>.</b>toHaveBeenCalledWithContext<d>(</b><g>expected</g><d>)</b>

Expected: not <g>Any<String></g>
Received:     <d>"foo"</b>

Number of calls: <r>1</g>
`;

exports[`toHaveBeenLastCalledWithContext includes the custom mock name in the error message 1`] = `
<d>expect(</b><r>named-mock</g><d>).</b>not<d>.</b>toHaveBeenLastCalledWithContext<d>(</b><g>expected</g><d>)</b>

Expected: not <g>"foo"</g>

Number of calls: <r>1</g>
`;

exports[`toHaveBeenLastCalledWithContext works only on spies or fn 1`] = `
<d>expect(</b><r>received</g><d>).</b>toHaveBeenLastCalledWithContext<d>(</b><g>expected</g><d>)</b>

<b>Matcher error</b>: <r>received</g> value must be a mock or spy function

Received has type:  function
Received has value: <r>[Function mock]</g>
`;

exports[`toHaveBeenLastCalledWithContext works when not called 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>toHaveBeenLastCalledWithContext<d>(</b><g>expected</g><d>)</b>

Expected: <g>"foo"</g>

Number of calls: <r>0</g>
`;

exports[`toHaveBeenLastCalledWithContext works with a context that doesn't match 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>toHaveBeenLastCalledWithContext<d>(</b><g>expected</g><d>)</b>

Expected: <g>"foo"</g>
Received: <r>"bar"</g>
Number of calls: <r>1</g>
`;

exports[`toHaveBeenLastCalledWithContext works with a context that doesn't match with matchers 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>toHaveBeenLastCalledWithContext<d>(</b><g>expected</g><d>)</b>

Expected: <g>Any<Number></g>
Received: <r>"bar"</g>
Number of calls: <r>1</g>
`;

exports[`toHaveBeenLastCalledWithContext works with a context that matches 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>not<d>.</b>toHaveBeenLastCalledWithContext<d>(</b><g>expected</g><d>)</b>

Expected: not <g>"foo"</g>

Number of calls: <r>1</g>
`;

exports[`toHaveBeenLastCalledWithContext works with a context that matches with a matcher 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>not<d>.</b>toHaveBeenLastCalledWithContext<d>(</b><g>expected</g><d>)</b>

Expected: not <g>Any<String></g>
Received:     <d>"foo"</b>

Number of calls: <r>1</g>
`;

exports[`toHaveBeenNthCalledWithContext includes the custom mock name in the error message 1`] = `
<d>expect(</b><r>named-mock</g><d>).</b>not<d>.</b>toHaveBeenNthCalledWithContext<d>(</b>n<d>, </b><g>expected</g><d>)</b>

n: 1
Expected: not <g>"foo"</g>

Number of calls: <r>1</g>
`;

exports[`toHaveBeenNthCalledWithContext works only on spies or fn 1`] = `
<d>expect(</b><r>received</g><d>).</b>toHaveBeenNthCalledWithContext<d>(</b>n<d>, </b><g>expected</g><d>)</b>

<b>Matcher error</b>: <r>received</g> value must be a mock or spy function

Received has type:  function
Received has value: <r>[Function mock]</g>
`;

exports[`toHaveBeenNthCalledWithContext works when not called 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>toHaveBeenNthCalledWithContext<d>(</b>n<d>, </b><g>expected</g><d>)</b>

n: 1
Expected: <g>"foo"</g>

Number of calls: <r>0</g>
`;

exports[`toHaveBeenNthCalledWithContext works with a context that doesn't match 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>toHaveBeenNthCalledWithContext<d>(</b>n<d>, </b><g>expected</g><d>)</b>

n: 1
Expected: <g>"foo"</g>
Received: <r>"bar"</g>
Number of calls: <r>1</g>
`;

exports[`toHaveBeenNthCalledWithContext works with a context that doesn't match with matchers 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>toHaveBeenNthCalledWithContext<d>(</b>n<d>, </b><g>expected</g><d>)</b>

n: 1
Expected: <g>Any<Number></g>
Received: <r>"bar"</g>
Number of calls: <r>1</g>
`;

exports[`toHaveBeenNthCalledWithContext works with a context that matches 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>not<d>.</b>toHaveBeenNthCalledWithContext<d>(</b>n<d>, </b><g>expected</g><d>)</b>

n: 1
Expected: not <g>"foo"</g>

Number of calls: <r>1</g>
`;

exports[`toHaveBeenNthCalledWithContext works with a context that matches with a matcher 1`] = `
<d>expect(</b><r>jest.fn()</g><d>).</b>not<d>.</b>toHaveBeenNthCalledWithContext<d>(</b>n<d>, </b><g>expected</g><d>)</b>

n: 1
Expected: not <g>Any<String></g>
Received:     <d>"foo"</b>

Number of calls: <r>1</g>
`;
